File: rust/src/redis/config/subscribe.rs
/// Configuration for Redis stream subscribers, including fetch count and block time.
///
/// Defaults:
/// - `consumer_name`: same as `topic_name`
/// - `group_name`: same as `topic_name`
/// - `topic_name`: as provided
/// - `num_fetch`: 10
/// - `block_time`: 5000 ms (5 seconds)
/// - `auto_claim`: 30000 ms (min-idle-time for xauto-claim)
#[derive(Clone, Debug)]
pub struct SubscriberConfig {
  pub(in super::super) consumer_name: String,
  pub(in super::super) group_name: String,
  pub(in super::super) topic_name: String,
  pub(in super::super) num_fetch: usize,
  pub(in super::super) block_time: usize,
  pub(in super::super) auto_claim: usize,
}

impl SubscriberConfig {
  /// Creates a new configuration with defaults and the given topic name.
  /// # Parameters
  ///
  /// * `topic_name` - The name of the Redis stream topic.
  ///
  pub fn new(topic_name: impl Into<String>) -> Self {
    let topic_name = topic_name.into();
    return Self {
      consumer_name: topic_name.clone(),
      group_name: topic_name.clone(),
      topic_name: topic_name,
      num_fetch: 10,     // Default number to fetch
      block_time: 5000,  // Default block time in milliseconds (5 seconds)
      auto_claim: 30000, // min-idle-time for xauto-claim in milliseconds (30 seconds)
    };
  }

  /// Sets the consumer name.
  pub fn consumer_name(mut self, consumer_name: impl Into<String>) -> Self {
    self.consumer_name = consumer_name.into();
    self
  }

  /// Sets the consumer group name.
  pub fn group_name(mut self, group_name: impl Into<String>) -> Self {
    self.group_name = group_name.into();
    self
  }

  /// Sets the topic (stream) name.
  pub fn topic_name(mut self, topic_name: impl Into<String>) -> Self {
    self.topic_name = topic_name.into();
    self
  }

  /// Sets how many messages to fetch per read.
  pub fn num_fetch(mut self, count: usize) -> Self {
    self.num_fetch = count;
    self
  }

  /// Sets the blocking time in milliseconds for each read.
  pub fn block_time(mut self, millis: usize) -> Self {
    self.block_time = millis;
    self
  }

  /// Sets the minimum idle time in milliseconds for auto-claiming pending messages.
  /// If the value is 0, auto-claiming is disabled.
  pub fn auto_claim(mut self, millis: usize) -> Self {
    self.auto_claim = millis;
    self
  }
}

File: go/brokers/redis/config/subscribe.go
package config

// SubscriberConfig holds the configuration settings needed to initialize and manage
// a Redis stream subscriber, including fetch count and block time.
//
// Default values:
//   - ConsumerName: same as TopicName
//   - GroupName: same as TopicName
//   - NumFetch: 10
//   - BlockTime: 5000 ms (5 seconds)
//   - AutoClaim: 30000 ms (min-idle-time for XAUTOCLAIM)
type SubscriberConfig struct {
	// ConsumerName is the name of the consumer within the consumer group.
	// If not set, it defaults to the stream name.
	ConsumerName string
	// GroupName is the Redis stream consumer group name.
	// If not set, it defaults to the stream name.
	GroupName string
	// TopicName is the name of the Redis stream topic.
	TopicName string
	// NumFetch is the number of messages to fetch per read.
	// Defaults to 10.
	NumFetch uint64
	// BlockTime is the blocking time in milliseconds for each read.
	// Defaults to 5000 (5 seconds).
	BlockTime uint64
	// AutoClaim is the minimum idle time in milliseconds for auto-claiming pending messages.
	// If set to 0, auto-claiming is disabled.
	// Defaults to 30000 (30 seconds).
	AutoClaim uint64
}

const (
	defaultNumFetch  = 10
	defaultBlockTime = 5000
	defaultAutoClaim = 30000
)

// NewSubscriberConfig creates a new SubscriberConfig instance with defaults and the given topic name.
//
// Default values:
//   - ConsumerName: same as topicName
//   - GroupName: same as topicName
//   - NumFetch: 10
//   - BlockTime: 5000 ms (5 seconds)
//   - AutoClaim: 30000 ms (30 seconds, min-idle-time for XAUTOCLAIM)
func NewSubscriberConfig(topicName string) *SubscriberConfig {
	return &SubscriberConfig{
		ConsumerName: topicName,
		GroupName:    topicName,
		TopicName:    topicName,
		NumFetch:     defaultNumFetch,
		BlockTime:    defaultBlockTime,
		AutoClaim:    defaultAutoClaim,
	}
}

// WithConsumerName sets the consumer name and returns the updated config.
// This method supports method chaining.
func (cfg *SubscriberConfig) WithConsumerName(consumerName string) *SubscriberConfig {
	cfg.ConsumerName = consumerName
	return cfg
}

// WithGroupName sets the consumer group name and returns the updated config.
// This method supports method chaining.
func (cfg *SubscriberConfig) WithGroupName(groupName string) *SubscriberConfig {
	cfg.GroupName = groupName
	return cfg
}

// WithTopicName sets the topic (stream) name and returns the updated config.
// This method supports method chaining.
func (cfg *SubscriberConfig) WithTopicName(topicName string) *SubscriberConfig {
	cfg.TopicName = topicName
	return cfg
}

// WithNumFetch sets how many messages to fetch per read and returns the updated config.
// This method supports method chaining.
func (cfg *SubscriberConfig) WithNumFetch(count uint64) *SubscriberConfig {
	cfg.NumFetch = count
	return cfg
}

// WithBlockTime sets the blocking time in milliseconds for each read and returns the updated config.
// This method supports method chaining.
func (cfg *SubscriberConfig) WithBlockTime(millis uint64) *SubscriberConfig {
	cfg.BlockTime = millis
	return cfg
}

// WithAutoClaim sets the minimum idle time in milliseconds for auto-claiming pending messages
// and returns the updated config. If the value is 0, auto-claiming is disabled.
// This method supports method chaining.
func (cfg *SubscriberConfig) WithAutoClaim(millis uint64) *SubscriberConfig {
	cfg.AutoClaim = millis
	return cfg
}
